### PR TITLE
Ab 936 incoming webhook ci

### PR DESCRIPTION
**Pull request by vivek-devtron**
# Description

User already has a CI running in some platform (Jenkins, Circle CI etc.) and they want to use Devtron for deployment by leveraging their existing CI

Fixes # [AB#936](https://dev.azure.com/DevtronLabs/Devtron/_workitems/edit/936)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?
- verified by creating new webhook CD
- deletion of webhook CD
- validated current flow working

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/common/helpers/workflowURL.ts [hunk 1]
```diff
@@ -1,9 +1,9 @@
 
-export function getCDPipelineURL(appId: string, workflowId: string, ciPipelineId: string, cdPipelineId: string = null) {
+export function getCDPipelineURL(appId: string, workflowId: string, ciPipelineId: string, cdPipelineId: string = null, isWebhookParent: boolean) {
```
Comment (sohel-rp) [inline src/components/common/helpers/workflowURL.ts:2]: Can we keep the optional arg at the end?

File: src/components/app/details/triggerView/workflow.service.ts [hunk 2]
```diff
@@ -80,13 +85,27 @@ export function processWorkflow(
                             return
                         }
                         wf.nodes.push(ciNode)
+                    } else if (branch.type == PipelineType.WEBHOOK) {
+                        let webhook = webhookMap.get(branch.componentId)
```
Comment (sohel-rp) [inline src/components/app/details/triggerView/workflow.service.ts:89]: Make it const

File: src/components/app/details/triggerView/workflow.service.ts [hunk 3]
```diff
@@ -80,13 +85,27 @@ export function processWorkflow(
...
                             return
                         }
                         let cdNode = cdPipelineToNode(cdPipeline, dimensions, branch.parentId)
                         let parentType = branch.parentType == PipelineType.CI_PIPELINE ? 'CI' : 'CD'
+                        if (branch.parentType == PipelineType.CI_PIPELINE) {
+                            parentType = 'CI'
+                        } else if (branch.parentType == PipelineType.WEBHOOK) {
+                            parentType = PipelineType.WEBHOOK
+                        } else {
+                            parentType = 'CD'
```
Comment (sohel-rp) [inline src/components/app/details/triggerView/workflow.service.ts:107]: Can we please remove else...if & else part? As we're already checking for CI/CD on Ln. 101.

File: src/components/app/details/triggerView/workflow.service.ts [hunk 4]
```diff
@@ -127,11 +146,16 @@ export function processWorkflow(
             s.width = dimensions.staticNodeSizes.nodeWidth
         })
 
-        ciNode.x =
-            startX +
-            workflowOffset.offsetX +
-            dimensions.staticNodeSizes.nodeWidth +
-            dimensions.staticNodeSizes.distanceX
+        if (ciNode.type === 'WEBHOOK') {
```
Comment (sohel-rp) [inline src/components/app/details/triggerView/workflow.service.ts:149]: Can we create const/enum type for CI/CD/WEBHOOK?

File: src/components/app/details/triggerView/workflow.service.ts [hunk 5]
```diff
@@ -305,6 +333,33 @@ function ciPipelineToNode(ciPipeline: CiPipeline, dimensions: WorkflowDimensions
     return ciNode
 }
 
+
+function webhookToNode(cdPipeline: CdPipeline, dimensions: WorkflowDimensions): NodeAttr {
+  let ciNode = {
```
Comment (sohel-rp) [inline src/components/app/details/triggerView/workflow.service.ts:338]: can be returned directly or changed to const?

File: src/components/app/details/triggerView/workflow/Workflow.tsx [hunk 6]
```diff
@@ -15,7 +16,9 @@ export class Workflow extends Component<WorkflowProps> {
                 return this.renderSourceNode(node)
             } else if (node.type === 'CI') {
                 return this.renderCINodes(node)
-            } else if (node.type === 'PRECD' || node.type === 'POSTCD') {
+            } else if (node.type === PipelineType.WEBHOOK) {
+              return this.renderWebhookNode(node)
+          } else if (node.type === 'PRECD' || node.type === 'POSTCD') {
```
Comment (sohel-rp) [inline src/components/app/details/triggerView/workflow/Workflow.tsx:21]: Also, include PRECD & POSTCD to the same const/enum type.

File: src/components/cdPipeline/CDPipeline.tsx [hunk 7]
```diff
@@ -1018,7 +1009,7 @@ export default class CDPipeline extends Component<CDPipelineProps, CDPipelineSta
                     </button>
                 )
             } else {
-                return (
+                return this.isWebhookCD? null :(
```
Comment (sohel-rp) [inline src/components/cdPipeline/CDPipeline.tsx:1012]: Change to `!this.isWebhookCD && (...JSX)`

File: src/components/cdPipeline/CDPipeline.tsx [hunk 8]
```diff
@@ -1095,10 +1086,56 @@ export default class CDPipeline extends Component<CDPipelineProps, CDPipelineSta
...
+    renderTriggerType() {
+      return (
+          <>
+              <p className="fs-14 fw-6 cn-9 mb-8">When do you want to deploy</p>
+              <div className="flex mb-20">
+                  <div
+                      className={`flex dc__content-start pointer w-50 pt-8 pr-16 pb-8 pl-16 br-4 mr-8 bw-1${
+                          this.state.pipelineConfig.triggerType === TriggerType.Auto ? ' bcb-1 eb-2' : ' bcn-0 en-2'
+                      }`}
+                      onClick={() => this.handleTriggerTypeChange(TriggerType.Auto)}
```
Comment (sohel-rp) [inline src/components/cdPipeline/CDPipeline.tsx:1103]: can we extract this to a separate function?

File: src/components/cdPipeline/CDPipeline.tsx [hunk 9]
```diff
@@ -1095,10 +1086,56 @@ export default class CDPipeline extends Component<CDPipelineProps, CDPipelineSta
...
+                      <div>
+                          <div>Automatic</div>
+                          <div>Deploy everytime a new image is received</div>
+                      </div>
+                  </div>
+                  <div
+                      className={`flex dc__content-start pointer w-50 pt-8 pr-16 pb-8 pl-16 br-4 ml-8 bw-1${
+                          this.state.pipelineConfig.triggerType === TriggerType.Manual ? ' bcb-1 eb-2' : ' bcn-0 en-2'
+                      }`}
+                      onClick={() => this.handleTriggerTypeChange(TriggerType.Manual)}
```
Comment (sohel-rp) [inline src/components/cdPipeline/CDPipeline.tsx:1115]: same here

Comment (sohel-rp) [reply src/components/cdPipeline/CDPipeline.tsx:1115]: data attr can be used to make the function more generic :)

File: src/components/ciPipeline/WebhookCD/WebhookDetails.tsx [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    return (
+        <Drawer position="right" width="1000px">
+            <div className="dc__window-bg h-100 webhook-details-container" ref={appStatusDetailRef}>
+                <div className="flex flex-align-center flex-justify dc__border-bottom bcn-0 pr-20">
+                    <h2 className="fs-16 fw-6 lh-1-43 m-0 title-padding">Deploy image from external source</h2>
+                    <button
+                        type="button"
+                        className="dc__transparent flex icon-dim-24"
+                        onClick={() => {
+                            close()
```
Comment (sohel-rp) [inline src/components/ciPipeline/WebhookCD/WebhookDetails.tsx:35]: same here

File: src/components/workflowEditor/PipelineSelect.tsx [hunk 11]
```diff
@@ -1,59 +1,77 @@
...
-                this.props.toggleCIMenu(event);
-                this.props.addCIPipeline('LINKED-CI', this.props.workflowId)
-            }}>
-                <div className="pipeline-webhook__icon">
-                    <img src={linkedPipeline} className="pipeline-webhook__icon" alt="linked-ci" />
+                <div
+                    className="flexbox pt-8 pr-12 pb-8 pl-12 pointer"
+                    onClick={(event) => {
+                        this.props.toggleCIMenu(event)
+                        this.props.addCIPipeline('CI', this.props.workflowId)
```
Comment (sohel-rp) [inline src/components/workflowEditor/PipelineSelect.tsx:19]: Same here. Extract to a different function.

File: src/components/workflowEditor/PipelineSelect.tsx [hunk 12]
```diff
@@ -1,59 +1,77 @@
...
+                    </div>
                 </div>
-                <div>
-                    <h4 className="pipeline-webhook__title">Linked CI Pipeline</h4>
-                    <p className="pipeline-webhook__text">Refer an existing Pipeline.</p>
+                <div
+                    className="flexbox pt-8 pr-12 pb-8 pl-12 pointer"
+                    onClick={(event) => {
+                        this.props.toggleCIMenu(event)
+                        this.props.addCIPipeline('LINKED-CI', this.props.workflowId)
```
Comment (sohel-rp) [inline src/components/workflowEditor/PipelineSelect.tsx:34]: same here

Comment (sohel-rp) [reply src/components/workflowEditor/PipelineSelect.tsx:34]: Again, We can make use of data attr :)

File: src/components/workflowEditor/PipelineSelect.tsx [hunk 13]
```diff
@@ -1,59 +1,77 @@
...
+                    Receive container image
                 </div>
-                <div>
-                    <h4 className="pipeline-webhook__title">Incoming Webhook</h4>
-                    <p className="pipeline-webhook__text">Receive container image from external source via a service.</p>
+                <div
+                    className="flexbox pt-8 pr-12 pb-8 pl-12 pointer"
+                    onClick={(event) => {
+                        this.props.toggleCIMenu(event)
+                        this.props.addWebhookCD(this.props.workflowId)
```
Comment (sohel-rp) [inline src/components/workflowEditor/PipelineSelect.tsx:52]: same here

File: src/components/common/helpers/workflowURL.ts (hunk 1 above)
Comment (vivek-devtron) [reply src/components/common/helpers/workflowURL.ts]: Done

File: src/components/app/details/triggerView/workflow.service.ts [hunk 14]
```diff
@@ -61,6 +63,9 @@ export function processWorkflow(
     let cdMap = new Map(
         (cdResponse?.pipelines ?? []).map((cdPipeline) => [cdPipeline.id, cdPipeline] as [number, CdPipeline]),
     )
+    let webhookMap = new Map(
```
[the remark this answers is not in the document]
Comment (vivek-devtron) [reply src/components/app/details/triggerView/workflow.service.ts]: done

File: src/components/app/details/triggerView/workflow.service.ts (hunk 3 above)
Comment (vivek-devtron) [reply src/components/app/details/triggerView/workflow.service.ts]: done

File: src/components/cdPipeline/CDPipeline.tsx (hunk 9 above)
Comment (vivek-devtron) [reply src/components/cdPipeline/CDPipeline.tsx]: Done

File: src/components/cdPipeline/CDPipeline.tsx (hunk 8 above)
Comment (vivek-devtron) [reply src/components/cdPipeline/CDPipeline.tsx]: Done

File: src/components/workflowEditor/PipelineSelect.tsx (hunk 13 above)
Comment (vivek-devtron) [reply src/components/workflowEditor/PipelineSelect.tsx]: Done

File: src/components/workflowEditor/PipelineSelect.tsx (hunk 12 above)
Comment (vivek-devtron) [reply src/components/workflowEditor/PipelineSelect.tsx]: Done

File: src/components/workflowEditor/PipelineSelect.tsx (hunk 11 above)
Comment (vivek-devtron) [reply src/components/workflowEditor/PipelineSelect.tsx]: done

File: src/components/ciPipeline/WebhookCD/WebhookDetails.tsx [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    useEffect(() => {
+        document.addEventListener('keydown', escKeyPressHandler)
+        return (): void => {
+            document.removeEventListener('keydown', escKeyPressHandler)
+        }
+    }, [escKeyPressHandler])
+    return (
+        <Drawer position="right" width="1000px">
```
[the remark this answers is not in the document]
Comment (vivek-devtron) [reply src/components/ciPipeline/WebhookCD/WebhookDetails.tsx]: done

File: src/components/cdPipeline/CDPipeline.tsx (hunk 7 above)
Comment (vivek-devtron) [reply src/components/cdPipeline/CDPipeline.tsx]: done

File: src/components/app/details/triggerView/workflow/Workflow.tsx (hunk 6 above)
Comment (vivek-devtron) [reply src/components/app/details/triggerView/workflow/Workflow.tsx:21]: done

File: src/components/app/details/triggerView/workflow.service.ts (hunk 4 above)
Comment (vivek-devtron) [reply src/components/app/details/triggerView/workflow.service.ts]: Done

File: src/components/app/details/triggerView/workflow/Workflow.tsx (hunk 6 above)
Comment (shivani170) [inline src/components/app/details/triggerView/workflow/Workflow.tsx:21]: Why aren't we using  constants for 'PRECD' and  'POSTCD'

File: src/components/app/details/triggerView/workflow/Workflow.tsx [hunk 16]
```diff
@@ -15,7 +16,9 @@ export class Workflow extends Component<WorkflowProps> {
...
                 return this.renderPrePostCDNodes(node)
             } else if (node.type === 'CD') {
```
Comment (shivani170) [inline src/components/app/details/triggerView/workflow/Workflow.tsx:23]: for 'ÇD' as well

File: src/components/ciPipeline/WebhookCD/WebhookDetails.tsx [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    useEffect(() => {
+        document.addEventListener('keydown', escKeyPressHandler)
+        return (): void => {
+            document.removeEventListener('keydown', escKeyPressHandler)
+        }
+    }, [escKeyPressHandler])
+    return (
+        <Drawer position="right" width="1024px">
```
Comment (shivani170) [inline src/components/ciPipeline/WebhookCD/WebhookDetails.tsx:31]: If this is a Drawer we can extend the name of the component as WebhookDetailsModal

File: src/components/workflowEditor/Workflow.tsx [hunk 18]
```diff
@@ -108,7 +111,8 @@ export class Workflow extends Component<WorkflowProps, WorkflowState> {
 
     renderNodes() {
         const ci = this.props.nodes.find((node) => node.type == 'CI')
-        const _nodesData = this.getNodesData(ci?.id || '')
+        const webhook = this.props.nodes.find((node) => node.type == 'WEBHOOK')
```
Comment (shivani170) [inline src/components/workflowEditor/Workflow.tsx:114]: Any constant for webhook

File: src/components/workflowEditor/nodes/WebhookNode.tsx [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+    triggerType: string
+    isLinkedCI: boolean
+    isExternalCI: boolean
+    isTrigger: boolean
+    linkedCount: number
+    to?: string
+    configDiffView?: boolean
+}
+
+export class WebhookNode extends Component<WebhookNodeProps> {
```
Comment (shivani170) [inline src/components/workflowEditor/nodes/WebhookNode.tsx:25]: Why class component?

File: src/components/workflowEditor/workflowEditor.css [hunk 20]
```diff
@@ -25,54 +25,12 @@
...
-
 .pipeline-webhook {
-    width: 416px;
-    padding: 0px 0px 16px 0px;
+    width: 320px;
+    padding: 0px 0px 12px 0px;
     box-shadow: 0 6px 12px -3px rgba(0, 56, 112, 0.3);
 }
 
-.pipeline-webhook__icon {
```
Comment (shivani170) [inline src/components/workflowEditor/workflowEditor.css:39]: This removal is intentional?

File: src/components/workflowEditor/nodes/WebhookNode.tsx (hunk 19 above)
Comment (vivek-devtron) [reply src/components/workflowEditor/nodes/WebhookNode.tsx]: done

File: src/components/workflowEditor/Workflow.tsx (hunk 18 above)
Comment (vivek-devtron) [reply src/components/workflowEditor/Workflow.tsx]: done

File: src/components/ciPipeline/WebhookCD/WebhookDetails.tsx (hunk 17 above)
Comment (vivek-devtron) [reply src/components/ciPipeline/WebhookCD/WebhookDetails.tsx:31]: Will do it in next PR

File: src/components/app/details/triggerView/workflow/Workflow.tsx (hunk 16 above)
Comment (vivek-devtron) [reply src/components/app/details/triggerView/workflow/Workflow.tsx]: Done

File: src/components/app/details/triggerView/workflow/Workflow.tsx (hunk 6 above)
Comment (vivek-devtron) [reply src/components/app/details/triggerView/workflow/Workflow.tsx]: Done

File: src/components/workflowEditor/workflowEditor.css (hunk 20 above)
Comment (vivek-devtron) [reply src/components/workflowEditor/workflowEditor.css:39]: yes

**Review by shivani170 — APPROVED** [review]

**Review by sohel-rp — APPROVED** [review]

**Pull request merged**